metrics/platform-metrics-reporting/src/main/java/ddf/metrics/reporting/internal/rrd4j/RrdMetricsRetriever.java

```diff
@@ -478,11 +496,131 @@ public OutputStream createXlsReport(List<String> metricNames, String metricsDir,
+            long startTime, long endTime, SUMMARY_INTERVALS summaryInterval)
+            throws IOException, MetricsGraphException {
+        // convert seconds to milliseconds
+        startTime *= MILLISECONDS_PER_SECOND;
+        endTime *= MILLISECONDS_PER_SECOND;
+        DateTime reportStart = new DateTime(startTime, DateTimeZone.UTC);
+        DateTime reportEnd = new DateTime(endTime, DateTimeZone.UTC);
+
+        Sheet sheet = wb.createSheet();
+        wb.setSheetName(0, reportStart.toString(SUMMARY_TIMESTAMP) + " to " + reportEnd
+                .toString(SUMMARY_TIMESTAMP));
+        Row headingRow = sheet.createRow(0);
+
+        int columnMax = 1;
+        for (String metricName : metricNames) {
+            MutableDateTime chunkStart = new MutableDateTime(reportStart);
+            MutableDateTime chunkEnd = new MutableDateTime(chunkStart);
+            Row row = sheet.createRow(metricNames.indexOf(metricName) + 1);
+            int columnCounter = 1;
+            Boolean isSum = null;
+
+            while (reportEnd.compareTo(chunkEnd) > 0 && columnCounter < EXCEL_MAX_COLUMNS) {
+                increment(chunkEnd, summaryInterval);
+                if (chunkEnd.isAfter(reportEnd)) {
+                    chunkEnd.setMillis(reportEnd);
+                }
+
+                // offset range by one millisecond so rrd will calculate granularity correctly
+                chunkEnd.addMillis(-1);
+                MetricData metricData = getMetricData(getRrdFilename(metricsDir, metricName),
+                        chunkStart.getMillis() / MILLISECONDS_PER_SECOND,
+                        chunkEnd.getMillis() / MILLISECONDS_PER_SECOND);
+                isSum = metricData.hasTotalCount();
+                chunkEnd.addMillis(1);
+
+                if (headingRow.getCell(columnCounter) == null) {
+                    Cell headingRowCell = headingRow.createCell(columnCounter);
+                    headingRowCell.getCellStyle().setWrapText(true);
+                    headingRowCell
+                            .setCellValue(getTimestamp(chunkStart, columnCounter, summaryInterval));
+                }
+
+                Cell sumOrAvg = row.createCell(columnCounter);
+                if (isSum) {
+                    sumOrAvg.setCellValue((double) metricData.getTotalCount());
+                } else {
+                    sumOrAvg.setCellValue(doubleAverage(metricData.getValues()));
+                }
+
+                chunkStart.setMillis(chunkEnd);
+                columnCounter++;
+            }
+            columnMax = columnCounter;
+
+            if (isSum != null) {
+                row.createCell(0).setCellValue(
+                        convertCamelCase(metricName) + " (" + (isSum ? "sum" : "avg") + ")");
+            }
+        }
+        for (int i = 0; i < columnMax; i++) {
+            sheet.autoSizeColumn(i);
+        }
+    }
+
+    private Double doubleAverage(Collection<Double> values) {
```

Comment:
It's an implementation of ![cumulative moving average](https://en.wikipedia.org/?title=Moving_average#Cumulative_moving_average) to avoid the edge case of a value overflow in the simple mean formula. I'm not sure how realistic that edge case is, but I did cause it by accident when fooling around with some sample data. Maybe Commons or Guava has a better implementation, I didn't see it, but I'll check again.